d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java

```diff
@@ -0,0 +1,217 @@
+import com.linkedin.d2.balancer.clients.TrackerClient;
+import com.linkedin.d2.balancer.strategies.StateUpdater;
+import com.linkedin.util.degrader.CallTracker;
+import com.linkedin.util.degrader.ErrorType;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+
+/**
+ * Update the state of the RelativeLoadBalancerStrategy periodically
+ */
+public class RelativeStateUpdater implements StateUpdater
+{
+  public static final double LOWEST_HEALTH_SCORE = 0.0;
+  public static final double HEALTHY_HEALTH_SCORE = 1.0;
+  // If slow start is enabled, it will always start from 0.01
+  private static final double SLOW_START_HEALTH_SCORE = 0.01;
+  private static final int SLOW_START_RECOVERY_FACTOR = 2;
+
+  private final D2RelativeStrategyProperties _relativeStrategyProperties;
+  private final QuarantineManager _quarantineManager;
+  private final ScheduledExecutorService _executorService;
+
+  // Keeps the state of each partition
+  private ConcurrentMap<Integer, PartitionLoadBalancerState> _partitionLoadBalancerStateMap;
+
+  RelativeStateUpdater(D2RelativeStrategyProperties relativeStrategyProperties,
+                       QuarantineManager quarantineManager,
+                       ScheduledExecutorService executorService)
+  {
+    _relativeStrategyProperties = relativeStrategyProperties;
+    _quarantineManager = quarantineManager;
+    _executorService = executorService;
+    _partitionLoadBalancerStateMap = new ConcurrentHashMap<>();
+
+    _executorService.schedule(this::updateState, _relativeStrategyProperties.getUpdateIntervalMs(), TimeUnit.MILLISECONDS);
+  }
+
+  @Override
+  public void updateStateByRequest(Set<TrackerClient> trackerClients, int partitionId, long clusterGenerationId)
+  {
+    if (!_partitionLoadBalancerStateMap.containsKey(partitionId))
+    {
+      PartitionLoadBalancerState partitionLoadBalancerState =  new PartitionLoadBalancerState(partitionId);
+      // If it is the very first request for the partition, update the state synchronously
+      if(partitionLoadBalancerState.getLock().tryLock())
+      {
+        try
+        {
+          _partitionLoadBalancerStateMap.put(partitionId,partitionLoadBalancerState);
+          updateStateForPartition(trackerClients, partitionId, Optional.of(clusterGenerationId));
+        }
+        finally
+        {
+          partitionLoadBalancerState.getLock().unlock();
+        }
+      }
+    } else if (_partitionLoadBalancerStateMap.get(partitionId).getClusterGenerationId() != clusterGenerationId)
+    {
+      // If there are uris change in the cluster, use executor to update the state asynchronously
+      _executorService.execute(() -> updateStateForPartition(trackerClients, partitionId, Optional.of(clusterGenerationId)));
```

Comment:
Should this be setting the TrackerClients in _partitionLoadBalancerStateMap instead of passing them directly?